hs_core/models.py

```diff
@@ -1508,41 +1512,205 @@ class Meta:
 
 
 class ResourceFile(models.Model):
     object_id = models.PositiveIntegerField()
     content_type = models.ForeignKey(ContentType)
-
     content_object = GenericForeignKey('content_type', 'object_id')
-    resource_file = models.FileField(upload_to=get_path, max_length=500, null=True, blank=True,
-                                     storage=IrodsStorage()
-                                     if getattr(settings, 'USE_IRODS', False)
-                                     else DefaultStorage())
+    # the file_folder attribute makes uploads work properly. It is otherwise not used.
+    file_folder = models.CharField(max_length=255, null=True)
+    resource_file = models.FileField(upload_to=get_path, max_length=500,
+                                     null=True, blank=True, storage=IrodsStorage())
     # the following optional fields are added for use by federated iRODS resources where
     # resources are created in the local federated zone rather than hydroshare zone, in
     # which case resource_file is empty, and we record iRODS logical resource file name
     # and file size for customized copy or move operations for resource creation and other
     # operations fed_resource_file in particular is a counterpart of resource_file, but
     # handles files uploaded from local disk and store the files to federated zone rather
     # than hydroshare zone.
-    fed_resource_file = models.FileField(upload_to=get_path, max_length=500, null=True, blank=True,
-                                         storage=IrodsStorage('federated') if getattr(
-                                             settings, 'USE_IRODS', False) else DefaultStorage())
+    fed_resource_file = models.FileField(upload_to=get_path, max_length=500,
+                                         null=True, blank=True, storage=IrodsStorage('federated'))
+    # DEPRECATED: utilize resfile.set_storage_path(path) and resfile.storage_path.
     fed_resource_file_name_or_path = models.CharField(max_length=255, null=True, blank=True)
+    # TODO: why is size a CharField???
     fed_resource_file_size = models.CharField(max_length=15, null=True, blank=True)
 
     @property
     def resource(self):
         return self.content_object
 
+    @property
+    def storage_path(self):
+        """
+        Return the qualified name for a file in the storage hierarchy.
+        This is a valid input to IrodsStorage for manipulating the file.
+        The output depends upon whether the IrodsStorage instance is running
+        in federated mode.
+
+        The heavy lifting in this routine is accomplished via get_path.
+        Regardless of whether the internal file name is qualified or not, get_path
+        makes it fully qualified.
+        """
+        if self.content_object.resource_federation_path:  # false if None or empty
+            return self.fed_resource_file.name
+        else:
+            return self.resource_file.name
```

Comment:
Even if ```resource_federation_path``` is not empty which indicates the resource is stored in a federated zone, ```fed_resource_file``` would be very likely be empty for majority of cases when a user selects a file from a federated zone using iRODS browser. In this case, ```fed_resource_file_name_or_path``` and ```fed_resource_file_size``` store the relative path in a federated zone where the resource is created, and there is no file upload action, hence ```fed_resource_file``` is empty. ```fed_resource_file``` is only set when a file from local disk is uploaded and added to an existing resource in a federated zone, or when the resource is initially created, the user selected a file from a federated zone using iRODS browser, then select another file from local disk, in which case, the resource will be created in the iRODS federated zone and fed_resource_file is set to associate with the file uploaded from local disk. Hence another conditional check is needed when ```resource_federation_path``` is true to see whether fed_resource_file_name_or_path is empty or not, and also checks whether fed_resource_file is empty or not, since they represent different cases and may be empty.